docs/WASI-design-principles.md

```diff
@@ -0,0 +1,95 @@
+
+ - Shared filesystem views - One of the unique capabilities WebAssembly brings to the table is the possibility of shared-nothing linking between applications and libraries. Shared-nothing means that all communication is via explicit calls, and the libraries don't share an address space or any other implicit shared state. But if we run both sides within the same filesystem view, that would give them a large body of shared state. Union mounts, mandatory access control systems, user namespaces, and other techniques can help, but often require complex configuration, heavy-weight boundaries, and sometimes even admin privileges to set up.
+
+This has wide-ranging implications. Much of POSIX is oriented around passing around strings, whether through command-line arguments, environment variables, or paths embedded in files, with the assumption that there's a shared filesystem view between components. As we said above, we're de-emphasizing strings, which dovetails with de-emphasizing shared filesystem views. Instead of having shared state and passing around values which identify things within the shared state, WASI prefers to share as little as possible, and use handles which represent the things which need to be shared.
+
+Compatibility with existing host environments and applications is important, and we have put a lot of work into WASI libc to provide POSIX compatibility and support existing applications. There's a lot more work that can be done, and a lot more we can do to improve compatibility and user convenience. We're continuing to make progress -- and users are encouraged to [file bugs](https://github.com/WebAssembly/WASI/issues) when they find things that don't work or are awkward. This approach supports existing applications, while also supporting applications and libraries willing to opt in to enable stronger and more fine-grained security properties than are possible in regular POSIX.
+
+For example, a typical POSIX-style API might include a function that accepts a file name to open. That requires the implementation to have a filesystem view, and to have appropriate permissions within that filesystem view. WASI APIs typically prefer to instead have a function which accepts a handle for an already-open file. That way, the implementation doesn't need a filesystem view, or permissions within the filesystem. It doesn't even need to care whether there even is a filesystem. When needed, compatibility with POSIX-style APIs is then provided as a thin layer on top implementing a simple name-to-handle mapping.
+
+We recognize that this approach has tradeoffs. It often does take more work to design and implement the compatibility layers needed to support existing applications than if we just always made WASI always expose POSIX-style APIs directly. And, even when we do have compatibility mechanisms, they aren't always the most locally optimal ones. The compatibility layer overhead is usually quite modest, but it is present.
```

Comment:
Double use of "always" in the second sentence seems odd.  Perhaps rephrase that sentence?    Do you need the "does" in "It often does take more work"?

Reply:
I cleaned up the "always". The "does" is for emphasis. This is is responding to certain people who publicly criticizing WASI over this; the intent is to acknowledge that there is a tradeoff being made here, and to argue that the tradeoff is worthwhile.